File: Sponsors/mw_streams_python/mws/operators/operator.py
from typing import List, TypeVar, Generic

from abc import ABC, abstractmethod


I = TypeVar("I")
O = TypeVar("O")


class Operator(ABC, Generic[I, O]):
    """
    An operator in the dataflow graph.
    """

    def __init__(self) -> None:
        self.children: List[Operator] = []
        ...

    @abstractmethod
    def consume(self, event: I) -> None:
        """
        Process some event and send the result downstream.
        """
        ...

    def add_child(self, child: "Operator") -> None:
        """
        Register a child (downstream node) to the operator.
        """
        self.children.append(child)

    def to(self, sink: "Sink") -> "Operator":
        self.add_child(sink)
        return sink

    def flush(self) -> None:
        """
        Emit any parially processed events (e.g. open windows) and
        flush downstream nodes.
        """
        for child in self.children:
            child.flush()

    def emit(self, event: O) -> None:
        for child in self.children:
            child.consume(event)


class Source(Operator[None, O]):
    """
    A source operator in a dataflow graph (one which produces values).
    """

    def __init__(self) -> None:
        super().__init__()

    @abstractmethod
    def start(self) -> None:
        """
        Begin producing values.
        """
        ...

    def consume(self, event) -> None:
        raise Exception(
            "source doesn't implement consume - no downstream nodes to read from"
        )


class Sink(Operator[I, None]):
    """
    A sink in a dataflow graph (a terminal node).
    """

    def __init__(self) -> None:
        super().__init__()

    @abstractmethod
    def consume(self, event: I) -> None: ...

    def add_child(self, child: Operator) -> None:
        raise Exception("a sink node cannot have children")

    def emit(self, event: None) -> None:
        raise Exception("a sink node cannot emit events")

File: Sponsors/mw-streams-java/src/main/java/com/mwam/hackathon/operators/Operator.java
package com.mwam.hackathon.operators;

import java.util.ArrayList;
import java.util.List;

/**
 * A node in a dataflow pipeline.
 *
 * @param <I>
 * @param <O>
 */
public abstract class Operator<I, O> {
    protected final List<Operator<O, ?>> children;

    protected Operator() {
        this.children = new ArrayList<>();
    }

    protected Operator(List<Operator<O, ?>> children) {
        this.children = children;
    }

    /**
     * Consumes some input and passes it to any children.
     *
     * @param input message to process
     */
    public abstract void consume(I input);

    /**
     * Register a child (downstream node) to the operator.
     *
     * @param child operator to pipe outputs to
     * @param <T>   output type of the child operator
     */
    public <T> void addChild(Operator<O, T> child) {
        this.children.add(child);
    }

    public void emit(O event) {
        for (var child : this.children) {
            child.consume(event);
        }
    }

    public void flush() {
        for (final var child : this.children) {
            child.flush();
        }
    }
}
